src/granite/base.cr

```diff
@@ -40,35 +40,46 @@ abstract class Granite::Base
   extend Select
 
   macro inherited
-    include JSON::Serializable
-    include YAML::Serializable
-
     @@select = Container.new(table_name: table_name, fields: fields)
 
     # Returns true if this object hasn't been saved yet.
-    @[JSON::Field(ignore: true)]
-    @[YAML::Field(ignore: true)]
     disable_granite_docs? property? new_record : Bool = true
 
     # Returns true if this object has been destroyed.
-    @[JSON::Field(ignore: true)]
-    @[YAML::Field(ignore: true)]
     disable_granite_docs? getter? destroyed : Bool = false
 
     # Returns true if the record is persisted.
     disable_granite_docs? def persisted?
       !(new_record? || destroyed?)
     end
 
-    disable_granite_docs? def initialize(**args : Granite::Columns::Type)
-      set_attributes(args.to_h.transform_keys(&.to_s))
+    # Consumes the result set to set self's property values.
+    disable_granite_docs? def initialize(result : DB::ResultSet) : Nil
+      {% verbatim do %}
+        {% begin %}
+          result.column_names.each do |col|
+            case col
+            {% for column in @type.instance_vars.select { |ivar| ivar.annotation(Granite::Column) } %}
+              {% ann = column.annotation(Granite::Column) %}
+              when {{column.name.stringify}} then @{{column.id}} = {% if ann[:converter] %} {{ann[:converter]}}.from_rs result {% else %} Granite::Type.from_rs(result, {{column.type}}) {% end %}
+            {% end %}
+            end
+          end
+        {% end %}
+      {% end %}
     end
 
-    disable_granite_docs? def initialize(args : Granite::ModelArgs)
-      set_attributes(args.transform_keys(&.to_s))
-    end
+    disable_granite_docs? def initialize(*args, **named_args)
```

Comment:
will this would work if someone defines their own transient properties?

Reply:
Oo no, might have to introduce an annotation or something to mark a property as "virtual".  The problem with just doing `@type.instance_vars` (without the annotation checks) would be it would try to set _any_ ivar.  Whether thats internal stuff like `@new_record`, ivars from other shards/modules etc.

Reply:
Why not just look over the list of fields we home we want to initialize? My memory is out of date but I recall a FIELDS constant that’s built by the individual field calls.

Reply:
The mutable constant stuff got removed when we switched to using annotations.